### PR TITLE
Favor decomposition and avoid matrix construction for large operations

### PR DESCRIPTION
**Pull request by AmintorDusko**
**Context:** PennyLane will attempt to construct a matrix for a given operation, if supported, instead of employing a decomposition. For large operations (e.g. `qml.QFT` with 6+ qubits), the matrix initialization times outweigh the time to apply its decomposition. For extensive operators, building a full matrix might not even be possible due to memory limits.

**Description of the Change:** A transition from dense matrix operation to the use of its decomposition is implemented. This transition is hard-coded for `QFT` and `Grover` operators, on the device level.

**Benefits:**
Faster calculations for large operators and a decrease in memory usage.
**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by AmintorDusko** [comment]
[sc-26765]

**Comment by AmintorDusko** [comment]
image

**Comment by AmintorDusko** [comment]
Regarding the QFT threshold defined here (6), this was the best setting on my laptop. 
This may depend on the architecture, let me know your thoughts.

File: pennylane/operation.py
Comment (albi3ro) [inline pennylane/operation.py:977]: Would this decomposition threshold vary depending on the device? Or would this hold for all simulators? If it's simulator-dependent, I would say this is information that the device would need to store, instead of the operation.

Even if it makes sense to be stored by the device, the only operators where this matters (so far) are `Permute` and `QFT`.  Maybe instead, it makes sense for the stopping condition to use `getattr(op, "decomposition_threshold", None)` and only define it in relevant classes.

File: pennylane/operation.py
Comment (albi3ro) [inline pennylane/operation.py:964]: We've started adding `has_` properties for the cases where:
a) sometimes we just want to know the existence of the property (like matrix)
b) it is expensive to compute by itself (like matrix)

I'd say it's fairly cheap to check `op.decomposition_threshold is not None`, so case b does not hold.

**Comment by josh146** [comment]
@AmintorDusko was discussing this with @albi3ro and @mlxd today; this feels like something that PennyLane core should not be deciding, but something that should be owned by the device.

Although, I believe that porting `lightning` away from inheriting from `default.qubit` should also address this issue at the same time, as `default.qubit` is the only device that supports matrices --- anything inheriting from `QubitDevice` will instead prefer decompositions.

**Review by albi3ro — COMMENTED** [review]
It would be good to add some unittests for `DefaultQubit.stopping_condition` for these cases in `tests/devices/test_default_qubit.py`.

**Review by mlxd — COMMENTED** [review]
Thanks @AmintorDusko 
We can pair this with a check on `lightning.qubit` and check the decomposition is used there also.
Happy to approve if so, since this will be mostly a functionality support change.

**Comment by AmintorDusko** [comment]
> Thanks @AmintorDusko We can pair this with a check on `lightning.qubit` and check the decomposition is used there also. Happy to approve if so, since this will be mostly a functionality support change.

Hey @mlxd, as it is now, `lightning.qubit` will inherit the `default.qubit` `stopping_condition` overload and the threshold functionality automatically kick in. Just to be sure, I tested it locally, the stopping_condition function works as it should for the `lightning.qubit`, and I was able to run comfortably a 24 wires QFT workload. I'm not able to run any workload with more-than-14-wires-full-matrix operators.

Changes need to be made in the [change device inheritance PR](https://github.com/PennyLaneAI/pennylane-lightning/pull/365), to transfer such functionality, though.
After we merge this PR, I will jump there and add the necessary changes, and tests. Does that look sensible?

File: pennylane/operation.py
Comment (AlbertMitjans) [reply pennylane/operation.py]: Why not generalising this to any operator?

For example, we could add a `use_decomposition` property to the needed operators that returns `True` if the number of wires used is higher than the `"decomposition_threshold"`. This way in the device we can just do:
```python
if getattr(obj, "use_decomposition", False):
    return obj.use_decomposition
```

File: doc/releases/changelog-dev.md
Comment (albi3ro) [inline doc/releases/changelog-dev.md:232]:
```suggestion
* `default.qubit` favours decomposition and avoids matrix construction for `QFT` and `GroverOperator` at larger qubit numbers.
```

File: pennylane/operation.py
Comment (AmintorDusko) [reply pennylane/operation.py]: Hi @AlbertMitjans, what you say makes sense.
The approach you propose is very similar to the first direction I follow.
But we decide to keep the decision fully owned by the device, and not as a property of the operation (in PL core).
For example on `lightning.qubit`, we may decide to choose different thresholds or even not use it at all, for some cases.
Since we only have a couple (2) cases, we decide to hard-code it for now.
There are no plans to expand this list in the near future.
Maybe @albi3ro would like to expand on that.

**Review by albi3ro — APPROVED** [review]
Left a suggestion to the changelog entry, but happy to approve otherwise.

File: doc/releases/changelog-dev.md
Comment (AmintorDusko) [reply doc/releases/changelog-dev.md:232]: Makes sense. Thank you for that!

**Comment by AmintorDusko** [comment]
> Left a suggestion to the changelog entry, but happy to approve otherwise.

Thank you, for all the help @albi3ro!

**Review by mlxd — APPROVED** [review]
Nothing else on my side. thanks @AmintorDusko

**Comment by AmintorDusko** [comment]
Thank you, @AlbertMitjans, @mlxd and @albi3ro!

**Pull request merged**
